kmem.c

```diff
@@ -0,0 +1,22 @@
+#include "kmem.h"
+#include "libkern.h"
+
+/* This is the pointer to the extended end of kernel's image. See
+   km_early_alloc for details. */
+static char* km_kernel_image_end;
+
+void km_init(){
+    /* The end of the kernel's .bss section. Provided by the linker. */
+    extern char __ebss[];
+
+    km_kernel_image_end = __ebss;
+}
+
+void* km_early_alloc(size_t size){
+    // TODO: Explicitly crash if this function is called after
+    // physical memory manager has been initialized.
+    void* ptr = km_kernel_image_end;
+    bzero(ptr, size);
+    km_kernel_image_end += size;
```

Comment:
MIPS CPU will trap on a word access to unaligned address. Please make sure that address returned by `km_early_alloc` is divisible by `sizeof(word_t)`.